test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionHelperTest.cs

```diff
@@ -0,0 +1,227 @@
+                    (Expression<Func<TestModel, string>>)(testModel => key),
+                    (Expression<Func<TestModel, TestModel>>)(m => m),
+                };
+            }
+        }
+
+        public static IEnumerable<object[]> IndexerExpressions
+        {
+            get
+            {
+                var i = 3;
+                var key = "TestModel";
+
+                return new TheoryData<Expression>
+                {
+                    (Expression<Func<IList<TestModel>, Category>>)(model => model[2].SelectedCategory),
+                    (Expression<Func<IList<TestModel>, CategoryName>>)(testModel => testModel[i].SelectedCategory.CategoryName),
+                    (Expression<Func<TestModel, int>>)(model => model.PreferredCategories[i].CategoryId),
+                    (Expression<Func<IDictionary<string, TestModel>, string>>)(model => model[key].SelectedCategory.CategoryName.MainCategory),
+                };
+            }
+        }
+
+        public static IEnumerable<object[]> IdenticalExpressions
+        {
+            get
+            {
+                var value = "Test";
+                var Model = "Test";
+                return new TheoryData<Expression, Expression>
+                {
+                    {
+                        (Expression<Func<TestModel, Category>>)(model => model.SelectedCategory),
+                        (Expression<Func<TestModel, Category>>)(model => model.SelectedCategory)
+                    },
+                    {
+                        (Expression<Func<TestModel, CategoryName>>)(model => model.SelectedCategory.CategoryName),
+                        (Expression<Func<TestModel, CategoryName>>)(model => model.SelectedCategory.CategoryName)
+                    },
+                    {
+                        (Expression<Func<TestModel, int>>)(testModel => testModel.SelectedCategory.CategoryId),
+                        (Expression<Func<TestModel, int>>)(testModel => testModel.SelectedCategory.CategoryId)
+                    },
+                    {
+                        (Expression<Func<TestModel, string>>)(model => model.SelectedCategory.CategoryName.MainCategory),
+                        (Expression<Func<TestModel, string>>)(model => model.SelectedCategory.CategoryName.MainCategory)
+                    },
+                    {
+                        (Expression<Func<TestModel, TestModel>>)(model => model),
+                        (Expression<Func<TestModel, TestModel>>)(m => m)
+                    },
+                    {
+                        (Expression<Func<TestModel, string>>)(model => value),
+                        (Expression<Func<TestModel, string>>)(m => value)
+                    },
+                    {
+                        (Expression<Func<TestModel, string>>)(m => Model),
+                        (Expression<Func<TestModel, TestModel>>)(m => m)
```

Comment:
Why do these two expressions use the same slot in the cache? Top-level nodes in the Linq expression should have different `NodeType`s.